### ISSUE TITLE
Memory leak when doing backward with grad as yourself

### ISSUE THREAD

**Issue by ebetica** [pytorch/pytorch#3818]
@zdevito  You said you wanted to make a python repro?

```
void backward(std::vector<Variable> outputs) {
   variable_list varlst;
   function_list funclst;
   for (auto v : outputs) {
     funclst.emplace_back(v.grad_fn(), v.output_nr());
     varlst.emplace_back(v);
   }
   detail::engine.execute(funclst, varlst, false);
}
```

This will leak a bunch of Tensors and in general be very sad for whatever Variables used.

**Comment by zdevito** [comment]
(context: it is unclear if this is just because we are running engine directly without checks, or whether this behavior can be reproduced in Python as well)

**Comment by apaszke** [comment]
Note that running backward like this (without making `outputs` volatile) will create a graph for backward too

**Comment by zdevito** [comment]
Yeah, that is fine. But when we all the tensors go out of scope the Function graph leaks...

**Comment by apaszke** [comment]
Just wanted to point out that if you run optimization using grads that have grad_fns, you might be accumulating history. A leak is quite unlikely considering that everything runs fine from Python.

**Comment by colesbury** [comment]
Here's the equivalent Python code (which also leaks):

```python
import gc
import resource
import torch
from torch.autograd import Variable

i = 0
while True:
    W = Variable(torch.randn(10), requires_grad=True)
    x = Variable(torch.randn(10), requires_grad=True)
    z = W * x
    z.backward(torch.ones(z.size()), create_graph=True)
    i += 1

    if i % 1000 == 0:
        gc.collect()
        max_mem_used = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
        print("{:.2f} MB".format(max_mem_used / 1024))
```

**Comment by colesbury** [comment]
So I think the cycle is:

`W` -> `[.grad]` -> `[.grad_fn]` -> `[.next_functions]` -> ... -> `grad_accumulator for W` -> `W`

Previously, we had the grad_accumulator have a weak reference to it's parent. I changed it when we switched Variable to ATen. We need to figure out a way to avoid the reference cycle.

**Comment by apaszke** [comment]
@colesbury is that possible? I thought we detach `grad_accumulator` from the variable when we make it non-leaf

**Comment by colesbury** [comment]
@apaszke, what do you mean by non-leaf? `W` is saved as a SavedVariable (and then unpcaked), but otherwise remains a leaf.

**Comment by apaszke** [comment]
Ok, never mind. That doesn't make sense. So we're probably leaking every time someone does 2nd order stuff (seems to align with some other reports)

**Comment by soumith** [comment]
to be clear this is only on master and not v0.3.0 correct?

**Comment by colesbury** [comment]
It's on v0.3.0 too. 2nd order stuff with `torch.autograd.grad` should be fine.

**Comment by dmarnerides** [comment]
What's the best way to help with this at the moment?

**Comment by colesbury** [comment]
@dmarnerides, are you seeing this problem? We'll have a fix soon. If you're encountering this, it'll be helpful to know if the upcoming PR fixes your issue.

**Comment by dmarnerides** [comment]
@colesbury I think it's likely that this is related to #3835 , which affects scenaria that require second derivatives, e.g. the gradient penalty for GANs. I think there were other recent reports on this too.

It would be great if there is a fix for this!

**Comment by apaszke** [comment]
Don't use `.backward(create_graph=True)`. Stick to `torch.autograd.grad` for now

**Comment by dmarnerides** [comment]
So just to make sure i get this correctly, you mean not use `.backward()` on the final loss (which might also contain terms evaluated with `torch.autograd.grad`) but instead compute that with `torch.autograd.grad` as well?

**Comment by apaszke** [comment]
Final loss is ok, because you don't give `create_graph=True`. It's that flag that causes the problems.

**Comment by dmarnerides** [comment]
But is it possible, without the flag, to compute the gradient w.r.t the input and then call backward() (second derivative) on that? Doesn't the final backward call require the graph of the first gradient?

**Comment by gchanan** [comment]
@skrah are you still looking at this?  I reproduced this on a recent master.

**Comment by rgommers** [comment]
@gchanan, @skrah will be back in December. PR gh-24368 addresses this issue and is close to done, if you're seeing this issue somewhere, perhaps test that PR?

**Comment by skrah** [comment]
@gchanan, I'll be back in January.  Perhaps I'll revisit the issue privately before then though. :-)

The general problem is that my PR (#24368)  currently fixes the leak but re-introduces some quirky semantics.

**Comment by kurtamohler** [comment]
Should someone take over? I think the fact that there's still an open PR that hasn't had any activity for more than a year might be dissuading people. I could try to take a stab at it--perhaps I could build off of #24368 and try to address the problems brought up there.

**Comment by ezyang** [comment]
@albanD just FYI that Quansight isn't going to try to finish the PR here, but that's not permanent, if you want to change your opinion on this.

**Comment by albanD** [comment]
That sounds good.
We added warning in code and doc to discourage people from doing backward(create_graph=True) so I don't think a fix for this is high priority. Especially because it would be a non-trivial change.
